Feature Request: Generate application properties by default

## Background

Currently, using Propactive requires the user to explicitly call  `:generateApplicationProperties` task to generate the `application.properties` file. It would be a better user experience if this is set as a sane default, with the option to disable. 

### Example:

```kotlin
propactive {
    implementationClass = "org.example.assignment.config.Properties"
    destination = layout.buildDirectory.dir("resources/main").get().asFile.absolutePath
    filenameOverride = "application.properties"
    // NEW
    autoGenerateApplicationProperties  = false // default is true
}
```

### Requirements

 - [x] Update README and task documentation to reflect the new feature.  
 - [x] Enable Plugin to configure a given build to generate the properties file by default.

I expect the E2E tests to be updated to reflect those changes.

### POC

You can see how this can be done on a KTS buildscript here:

```kotlin
propactive {
    implementationClass = "org.example.config.Properties"
    // TODO: There won't be a need to add this by default after #91 is resolved.
    classCompileDependency = "compileKotlin"
}

tasks {
    // TODO: There won't be a need to add this by default after #91 is resolved.
    named("classes") {
        dependsOn("generateApplicationProperties")
    }
}
```

Which means we might need to set `classCompileDependancy`  to something other than `classes` (unless you figured a way to create a valid chain of build dependency that will allow you to depend on `classes` by default, and yet also use it (`classes`) to run a `":generateApplicationProperties"` task afterwards)

To expand, it's not possible (also discouraged) to finalize on  `classes` as subsequent tasks cannot automatically declare an explicit or implicit dependency on output of task `:generateApplicationProperties`

```
FAILURE: Build failed with an exception.

* What went wrong:
A problem was found with the configuration of task ':resolveMainClassName' (type 'ResolveMainClassName').
  - Gradle detected a problem with the following location: '$HOME/workspace/github/JBR-HAX/jb-rsch-4744-u-ways/build/resources/main'.
    
    Reason: Task ':resolveMainClassName' uses this output of task ':generateApplicationProperties' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
    
    Possible solutions:
      1. Declare task ':generateApplicationProperties' as an input of ':resolveMainClassName'.
      2. Declare an explicit dependency on ':generateApplicationProperties' from ':resolveMainClassName' using Task#dependsOn.
      3. Declare an explicit dependency on ':generateApplicationProperties' from ':resolveMainClassName' using Task#mustRunAfter.
    
    For more information, please refer to https://docs.gradle.org/8.6/userguide/validation_problems.html#implicit_dependency in the Gradle documentation.
```

**Idea**: `classCompileDependancy` default should be `compileJava` (if found) and `compileKotlin` (if found). Which will then open the possibility of allowing to safely attach a `generateApplicationProperties` dependency for `classes`.

___

--- a/README.md
+++ b/README.md
@@ -78,24 +78,49 @@ propactive {
 }
 ```
 
-### Optional: Class Compile Optimisation
-
-Since Propactive is a runtime property generator that relies on loading a properties class at runtime, we need to ensure that
-the properties class is compiled before the `generateApplicationProperties` task is executed. By default, the plugin will set
-the `classCompileDependency` option to [`classes`](https://docs.gradle.org/current/userguide/java_plugin.html#sec:java_tasks),
-which is an aggregate task that just depends on other tasks in which other plugins usually attach additional compilation tasks 
-to it.
-
-However, you can set the `classCompileDependency` option to `compileJava` (or `compileKotlin` if you are using Kotlin) if you
-want to optimise your build time. As doing so will only regenerate the properties file when the properties class is recompiled:
+The default values for the `propactive` extension are as follows:
 
 ```kotlin
 propactive {
-    classCompileDependency = "compileKotlin"
+    environments = "*"
+    implementationClass = "ApplicationProperties"
+    destination = layout.buildDirectory.dir("resources/main").get().asFile.absolutePath
+    autoGenerateApplicationProperties = true
+    filenameOverride = null
+    classCompileDependency = null
+}
+```
+
+### Optional: Enable Class Compile Optimisation For Custom Compilation Tasks
+
+Since Propactive is a runtime property generator that relies on loading a properties class at runtime, we need to ensure that
+the properties class is compiled before the `generateApplicationProperties` task is executed. By default, the plugin will set
+the `classCompileDependency` option to `compileJava` or/and `compileKotlin` if you are using Kotlin.
+
+However, you can set the `classCompileDependency` option to something else if you want to optimise your build time, or if
+you are compiling your classes from a different source set. For example, if you are [using Scala](https://xebia.com/blog/scala-and-kotlin-under-one-roof), 
+you can set the `classCompileDependency` option to `compileScala`:
+
+```kotlin
+propactive {
+    classCompileDependency = "compileScala"
+}
+```
+
+### Optional: Disable Auto-Generated Properties File
+
+By default, the plugin will generate the properties file when the `classes` task is executed. If you want to disable this
+behaviour, you can set the `autoGenerateApplicationProperties` option to `false`:
+
+```kotlin
+propactive {
+    autoGenerateApplicationProperties = false
 }
 ```
 
 ## Plugin Tasks
+
+Propactive provides 2 tasks that you can use to generate and validate your application properties files:
 
 ```
 Propactive tasks

--- a/propactive-plugin/src/main/kotlin/io/github/propactive/plugin/Configuration.kt
+++ b/propactive-plugin/src/main/kotlin/io/github/propactive/plugin/Configuration.kt
@@ -10,6 +10,7 @@ import org.gradle.api.tasks.Input
  * @param destination Sets the location of your generated properties file within the build directory.
  * @param filenameOverride Allows overriding given filename for an environment. (This can only be used when generating application properties for a singular environment)
  * @param classCompileDependency The task Propactive will depend on to load/compile your properties class/object you created.
+ * @param autoGenerateApplicationProperties Automatically generate application properties file for each given environment.
  */
 open class Configuration(
     @get:Input var environments: String = DEFAULT_ENVIRONMENTS,
@@ -17,15 +18,17 @@ open class Configuration(
     @get:Input var destination: String = DEFAULT_BUILD_DESTINATION,
     @get:Input var filenameOverride: String = DEFAULT_FILENAME_OVERRIDE,
     @get:Input var classCompileDependency: String = DEFAULT_CLASS_COMPILE_DEPENDENCY,
+    @get:Input var autoGenerateApplicationProperties: Boolean = DEFAULT_AUTO_GENERATE_APPLICATION_PROPERTIES,
 ) {
     companion object {
         // Cached default configuration
         internal val DEFAULT_CONFIGURATION = Configuration()
         internal const val DEFAULT_ENVIRONMENTS = "*"
         internal const val DEFAULT_IMPLEMENTATION_CLASS = "ApplicationProperties"
-        internal const val DEFAULT_CLASS_COMPILE_DEPENDENCY = "classes"
+        internal const val DEFAULT_AUTO_GENERATE_APPLICATION_PROPERTIES = true
 
         // NOTE: Gradle plugin extension values cannot be null if you want to register them as property values
+        internal const val DEFAULT_CLASS_COMPILE_DEPENDENCY = ""
         internal const val DEFAULT_BUILD_DESTINATION = ""
         internal const val DEFAULT_FILENAME_OVERRIDE = ""
     }

--- a/propactive-plugin/src/main/kotlin/io/github/propactive/plugin/Propactive.kt
+++ b/propactive-plugin/src/main/kotlin/io/github/propactive/plugin/Propactive.kt
@@ -14,45 +14,79 @@ open class Propactive : Plugin<Project> {
             .apply { LOGGER.debug("Creating extension: {}", Configuration::class.simpleName) }
             .create(PROPACTIVE_GROUP, Configuration::class.java)
 
-        target
+        val applicationPropertiesTaskTaskProvider = target
             .tasks
             .apply { LOGGER.debug("Registering Task: {}", GenerateApplicationPropertiesTask.TASK_NAME) }
             .register(GenerateApplicationPropertiesTask.TASK_NAME, GenerateApplicationPropertiesTask::class.java)
 
-        target
+        val validateApplicationPropertiesTaskTaskProvider = target
             .tasks
             .apply { LOGGER.debug("Registering Task: {}", ValidateApplicationPropertiesTask.TASK_NAME) }
             .register(ValidateApplicationPropertiesTask.TASK_NAME, ValidateApplicationPropertiesTask::class.java)
 
-        target
-            .extensions
-            .apply { LOGGER.debug("Configuring extension: {}", Configuration::class.simpleName) }
-            .findByType(Configuration::class.java)
-            ?.apply {
-                fun String.debug(property: KMutableProperty1<Configuration, String>) = apply {
-                    LOGGER.debug("Configuring '{}' with value: {}", property.name, this)
+        /** After evaluate ensures that the code within the block is executed after the project's configuration phase is complete */
+        target.afterEvaluate {
+            target
+                .extensions
+                .apply { LOGGER.debug("Evaluating $PROPACTIVE_GROUP extension") }
+                .findByType(Configuration::class.java)
+                ?.apply {
+                    fun <T> T.debug(property: KMutableProperty1<Configuration, T>) = apply {
+                        LOGGER.debug("Setting '{}' with value: {}", property.name, this)
+                    }
+
+                    environments = target
+                        .propertyOrDefault(Configuration::environments.name, environments)
+                        .debug(Configuration::environments)
+
+                    implementationClass = target
+                        .propertyOrDefault(Configuration::implementationClass.name, implementationClass)
+                        .debug(Configuration::implementationClass)
+
+                    destination = target
+                        .propertyOrDefault(Configuration::destination.name, destination)
+                        .debug(Configuration::destination)
+
+                    filenameOverride = target
+                        .propertyOrDefault(Configuration::filenameOverride.name, filenameOverride)
+                        .debug(Configuration::filenameOverride)
+
+                    classCompileDependency = target
+                        .propertyOrDefault(Configuration::classCompileDependency.name, classCompileDependency)
+                        .debug(Configuration::classCompileDependency)
+                        .also { dependency ->
+                            /**
+                             * We need to ensure the task depends on the compile task to ensure the application properties
+                             * class is loaded before we attempt to generate the application properties file. Therefore,
+                             * if the user has not provided a custom task to depend on, we'll default to the compileKotlin,
+                             * compileJava, or both tasks.
+                             */
+                            (
+                                dependency.takeUnless(String::isBlank)
+                                    ?.let(::listOf)
+                                    ?: listOfNotNull(
+                                        target.tasks.findByName("compileKotlin"),
+                                        target.tasks.findByName("compileJava"),
+                                    )
+                                )
+                                .forEach { task ->
+                                    applicationPropertiesTaskTaskProvider.get().dependsOn(task)
+                                    validateApplicationPropertiesTaskTaskProvider.get().dependsOn(task)
+                                }
+                        }
+
+                    /** This configuration does not support setters via the system properties. */
+                    autoGenerateApplicationProperties
+                        .debug(Configuration::autoGenerateApplicationProperties)
+                        .let { isTrue ->
+                            if (isTrue) {
+                                target.tasks.named("classes") { task ->
+                                    task.dependsOn.add(GenerateApplicationPropertiesTask.TASK_NAME)
+                                }
+                            }
+                        }
                 }
-
-                environments = target
-                    .propertyOrDefault(Configuration::environments.name, environments)
-                    .debug(Configuration::environments)
-
-                implementationClass = target
-                    .propertyOrDefault(Configuration::implementationClass.name, implementationClass)
-                    .debug(Configuration::implementationClass)
-
-                destination = target
-                    .propertyOrDefault(Configuration::destination.name, destination)
-                    .debug(Configuration::destination)
-
-                filenameOverride = target
-                    .propertyOrDefault(Configuration::filenameOverride.name, filenameOverride)
-                    .debug(Configuration::filenameOverride)
-
-                classCompileDependency = target
-                    .propertyOrDefault(Configuration::classCompileDependency.name, classCompileDependency)
-                    .debug(Configuration::classCompileDependency)
-            }
+        }
     }
 
     companion object {

--- a/propactive-plugin/src/main/kotlin/io/github/propactive/task/ApplicationPropertiesTask.kt
+++ b/propactive-plugin/src/main/kotlin/io/github/propactive/task/ApplicationPropertiesTask.kt
@@ -75,6 +75,5 @@ abstract class ApplicationPropertiesTask : DefaultTask() {
 
     init {
         group = Propactive.PROPACTIVE_GROUP
-        dependsOn.add(configuration.classCompileDependency)
     }
 }

--- a/propactive-plugin/src/test/kotlin/io/github/propactive/matcher/ConfigurationMatcher.kt
+++ b/propactive-plugin/src/test/kotlin/io/github/propactive/matcher/ConfigurationMatcher.kt
@@ -8,7 +8,7 @@ import io.kotest.matchers.equalityMatcher
 import io.kotest.matchers.should
 
 class ConfigurationMatcher private constructor() : Matcher<Configuration> {
-    private val matchers: MutableMap<Field, Matcher<String>> = mutableMapOf()
+    private val matchers: MutableMap<Field, Matcher<Any>> = mutableMapOf()
 
     companion object {
         fun Configuration.shouldMatch(matcher: ConfigurationMatcher.() -> Unit) =
@@ -35,6 +35,10 @@ class ConfigurationMatcher private constructor() : Matcher<Configuration> {
         matchers[Field.IMPLEMENTATION_CLASS_COMPILE_DEPENDENCY] = equalityMatcher(expected)
     }
 
+    internal fun withAutoGenerateApplicationProperties(expected: Boolean) = apply {
+        matchers[Field.AUTO_GENERATE_APPLICATION_PROPERTIES] = equalityMatcher(expected)
+    }
+
     override fun test(value: Configuration): MatcherResult = matchers
         // All Configuration fields should be matched...
         .apply { keys.shouldContainAll(Field.entries.toSet()) }
@@ -52,11 +56,12 @@ class ConfigurationMatcher private constructor() : Matcher<Configuration> {
         )
     }
 
-    private enum class Field(val extractor: (Configuration) -> String) {
+    private enum class Field(val extractor: (Configuration) -> Any) {
         IMPLEMENTATION_CLASS(Configuration::implementationClass),
         DESTINATION(Configuration::destination),
         ENVIRONMENTS(Configuration::environments),
         FILENAME_OVERRIDE(Configuration::filenameOverride),
         IMPLEMENTATION_CLASS_COMPILE_DEPENDENCY(Configuration::classCompileDependency),
+        AUTO_GENERATE_APPLICATION_PROPERTIES(Configuration::autoGenerateApplicationProperties),
     }
 }

--- a/propactive-plugin/src/test/kotlin/io/github/propactive/plugin/ConfigurationTest.kt
+++ b/propactive-plugin/src/test/kotlin/io/github/propactive/plugin/ConfigurationTest.kt
@@ -1,14 +1,16 @@
 package io.github.propactive.plugin
 
 import io.github.propactive.matcher.ConfigurationMatcher.Companion.shouldMatch
+import io.github.propactive.plugin.Configuration.Companion.DEFAULT_AUTO_GENERATE_APPLICATION_PROPERTIES
 import io.github.propactive.plugin.Configuration.Companion.DEFAULT_BUILD_DESTINATION
+import io.github.propactive.plugin.Configuration.Companion.DEFAULT_CLASS_COMPILE_DEPENDENCY
 import io.github.propactive.plugin.Configuration.Companion.DEFAULT_ENVIRONMENTS
 import io.github.propactive.plugin.Configuration.Companion.DEFAULT_FILENAME_OVERRIDE
 import io.github.propactive.plugin.Configuration.Companion.DEFAULT_IMPLEMENTATION_CLASS
-import io.github.propactive.plugin.Configuration.Companion.DEFAULT_CLASS_COMPILE_DEPENDENCY
 import io.kotest.matchers.nulls.shouldNotBeNull
 import org.junit.jupiter.api.Test
 import java.util.UUID.randomUUID
+import kotlin.random.Random
 
 internal class ConfigurationTest {
 
@@ -21,6 +23,7 @@ internal class ConfigurationTest {
                 withFilenameOverride(DEFAULT_FILENAME_OVERRIDE)
                 withEnvironments(DEFAULT_ENVIRONMENTS)
                 withImplementationClassCompileDependency(DEFAULT_CLASS_COMPILE_DEPENDENCY)
+                withAutoGenerateApplicationProperties(DEFAULT_AUTO_GENERATE_APPLICATION_PROPERTIES)
             }
     }
 
@@ -33,6 +36,7 @@ internal class ConfigurationTest {
                 withFilenameOverride(DEFAULT_FILENAME_OVERRIDE)
                 withEnvironments(DEFAULT_ENVIRONMENTS)
                 withImplementationClassCompileDependency(DEFAULT_CLASS_COMPILE_DEPENDENCY)
+                withAutoGenerateApplicationProperties(DEFAULT_AUTO_GENERATE_APPLICATION_PROPERTIES)
             }
     }
 
@@ -43,6 +47,7 @@ internal class ConfigurationTest {
         val destinationNewValue = "${randomUUID()}"
         val filenameOverrideNewValue = "${randomUUID()}"
         val implementationClassCompileDependencyNewValue = "${randomUUID()}"
+        val autoGenerateApplicationPropertiesNewValue = Random.nextBoolean()
 
         Configuration()
             .apply {
@@ -51,24 +56,28 @@ internal class ConfigurationTest {
                 destination = "${randomUUID()}"
                 filenameOverride = "${randomUUID()}"
                 classCompileDependency = "${randomUUID()}"
+                autoGenerateApplicationProperties = Random.nextBoolean()
             }.apply {
                 environments.shouldNotBeNull()
                 implementationClass.shouldNotBeNull()
                 destination.shouldNotBeNull()
                 filenameOverride.shouldNotBeNull()
                 classCompileDependency.shouldNotBeNull()
+                autoGenerateApplicationProperties.shouldNotBeNull()
             }.apply {
                 environments = environmentsNewValue
                 implementationClass = implementationClassNewValue
                 destination = destinationNewValue
                 filenameOverride = filenameOverrideNewValue
                 classCompileDependency = implementationClassCompileDependencyNewValue
+                autoGenerateApplicationProperties = autoGenerateApplicationPropertiesNewValue
             }.shouldMatch {
                 withEnvironments(environmentsNewValue)
                 withImplementationClass(implementationClassNewValue)
                 withDestination(destinationNewValue)
                 withFilenameOverride(filenameOverrideNewValue)
                 withImplementationClassCompileDependency(implementationClassCompileDependencyNewValue)
+                withAutoGenerateApplicationProperties(autoGenerateApplicationPropertiesNewValue)
             }
     }
 }

--- a/propactive-plugin/src/test/kotlin/io/github/propactive/plugin/PropactiveIT.kt
+++ b/propactive-plugin/src/test/kotlin/io/github/propactive/plugin/PropactiveIT.kt
@@ -4,6 +4,7 @@ import io.github.propactive.support.extension.KotlinEnvironmentExtension
 import io.github.propactive.support.extension.PublishSnapshotJars
 import io.github.propactive.support.extension.gradle.TaskExecutor
 import io.github.propactive.support.extension.gradle.TaskExecutor.Outcome
+import io.github.propactive.support.extension.project.ProjectDirectory
 import io.github.propactive.task.GenerateApplicationPropertiesTask
 import io.github.propactive.task.ValidateApplicationPropertiesTask
 import io.kotest.assertions.throwables.shouldNotThrow
@@ -30,13 +31,14 @@ class PropactiveIT {
     ) {
         shouldNotThrow<UnexpectedBuildFailure> {
             taskExecutor
-                .execute("assemble")
+                .execute("init")
         }
     }
 
     @Test
     @Order(2)
-    fun `should display Propactive's tasks description`(
+    fun `should register propactive tasks`(
+        projectDirectory: ProjectDirectory,
         taskExecutor: TaskExecutor,
     ) {
         taskExecutor

--- a/propactive-plugin/src/test/kotlin/io/github/propactive/plugin/PropactiveTest.kt
+++ b/propactive-plugin/src/test/kotlin/io/github/propactive/plugin/PropactiveTest.kt
@@ -1,6 +1,7 @@
 package io.github.propactive.plugin
 
 import io.github.propactive.matcher.ConfigurationMatcher.Companion.shouldMatch
+import io.github.propactive.plugin.Configuration.Companion.DEFAULT_AUTO_GENERATE_APPLICATION_PROPERTIES
 import io.github.propactive.plugin.Configuration.Companion.DEFAULT_BUILD_DESTINATION
 import io.github.propactive.plugin.Configuration.Companion.DEFAULT_CLASS_COMPILE_DEPENDENCY
 import io.github.propactive.plugin.Configuration.Companion.DEFAULT_ENVIRONMENTS
@@ -135,16 +136,18 @@ class PropactiveTest {
                     withDestination(DEFAULT_BUILD_DESTINATION)
                     withFilenameOverride(DEFAULT_FILENAME_OVERRIDE)
                     withImplementationClassCompileDependency(DEFAULT_CLASS_COMPILE_DEPENDENCY)
+                    withAutoGenerateApplicationProperties(DEFAULT_AUTO_GENERATE_APPLICATION_PROPERTIES)
                 }
         }
 
         @Test
-        fun `should allow modifying default configurations`() {
+        fun `should allow modifying default configurations via propactive extension`() {
             val customEnvironments = Random.alphaNumeric("customEnvironments")
             val customImplementationClass = Random.alphaNumeric("customImplementationClass")
             val customDestination = Random.alphaNumeric("customDestination")
             val customFilenameOverride = Random.alphaNumeric("customFilenameOverride")
             val customImplementationClassCompileDependency = Random.alphaNumeric("customImplementationClassCompileDependency")
+            val customAutoGenerateApplicationProperties = Random.nextBoolean()
 
             project
                 .extensions
@@ -155,6 +158,7 @@ class PropactiveTest {
                     destination = customDestination
                     filenameOverride = customFilenameOverride
                     classCompileDependency = customImplementationClassCompileDependency
+                    autoGenerateApplicationProperties = customAutoGenerateApplicationProperties
                 }
 
             project
@@ -166,11 +170,12 @@ class PropactiveTest {
                     withDestination(customDestination)
                     withFilenameOverride(customFilenameOverride)
                     withImplementationClassCompileDependency(customImplementationClassCompileDependency)
+                    withAutoGenerateApplicationProperties(customAutoGenerateApplicationProperties)
                 }
         }
 
         @Test
-        fun `should allow setting propactive configurations through system properties`() {
+        fun `should allow setting propactive configurations through system properties during runtime`() {
             val customEnvironments = Random.alphaNumeric("customEnvironments")
             val customImplementationClass = Random.alphaNumeric("customImplementationClass")
             val customDestination = Random.alphaNumeric("customDestination")
@@ -209,11 +214,12 @@ class PropactiveTest {
                     withDestination(customDestination)
                     withFilenameOverride(customFilenameOverride)
                     withImplementationClassCompileDependency(customImplementationClassCompileDependency)
+                    withAutoGenerateApplicationProperties(DEFAULT_AUTO_GENERATE_APPLICATION_PROPERTIES)
                 }
         }
 
         @Test
-        fun `should allow overriding propactive custom configurations with system properties`() {
+        fun `should allow overriding propactive custom configurations with system properties during runtime`() {
             val customConfigEnvironments = Random.alphaNumeric("customConfigEnvironments")
             val customConfigImplementationClass = Random.alphaNumeric("customConfigImplementationClass")
             val customConfigDestination = Random.alphaNumeric("customConfigDestination")
@@ -248,6 +254,7 @@ class PropactiveTest {
                     withDestination(customConfigDestination)
                     withFilenameOverride(customConfigFilenameOverride)
                     withImplementationClassCompileDependency(customConfigImplementationClassCompileDependency)
+                    withAutoGenerateApplicationProperties(DEFAULT_AUTO_GENERATE_APPLICATION_PROPERTIES)
                 }
 
             val properties = java.util.Properties().apply {
@@ -282,6 +289,7 @@ class PropactiveTest {
                     withDestination(customPropertyDestination)
                     withFilenameOverride(customPropertyFilenameOverride)
                     withImplementationClassCompileDependency(customPropertyImplementationClassCompileDependency)
+                    withAutoGenerateApplicationProperties(DEFAULT_AUTO_GENERATE_APPLICATION_PROPERTIES)
                 }
         }
     }

--- a/propactive-plugin/src/test/kotlin/io/github/propactive/support/extension/project/BuildScript.kt
+++ b/propactive-plugin/src/test/kotlin/io/github/propactive/support/extension/project/BuildScript.kt
@@ -1,5 +1,6 @@
 package io.github.propactive.support.extension.project
 
+import io.github.propactive.plugin.Configuration.Companion.DEFAULT_AUTO_GENERATE_APPLICATION_PROPERTIES
 import io.github.propactive.plugin.Configuration.Companion.DEFAULT_BUILD_DESTINATION
 import io.github.propactive.plugin.Configuration.Companion.DEFAULT_CLASS_COMPILE_DEPENDENCY
 import io.github.propactive.plugin.Configuration.Companion.DEFAULT_ENVIRONMENTS
@@ -17,6 +18,7 @@ class BuildScript(parent: File, name: String) : File(parent, name) {
         destination: String = DEFAULT_BUILD_DESTINATION,
         filenameOverride: String = DEFAULT_FILENAME_OVERRIDE,
         classCompileDependency: String = DEFAULT_CLASS_COMPILE_DEPENDENCY,
+        autoGenerateApplicationProperties: Boolean = DEFAULT_AUTO_GENERATE_APPLICATION_PROPERTIES,
     ) = apply {
         writeText(
             """
@@ -35,6 +37,7 @@ class BuildScript(parent: File, name: String) : File(parent, name) {
                 destination = "$destination"
                 filenameOverride = "$filenameOverride"
                 classCompileDependency = "$classCompileDependency"
+                autoGenerateApplicationProperties = $autoGenerateApplicationProperties
             }
 
             repositories {

--- a/propactive-plugin/src/test/kotlin/io/github/propactive/task/GenerateApplicationPropertiesTaskIT.kt
+++ b/propactive-plugin/src/test/kotlin/io/github/propactive/task/GenerateApplicationPropertiesTaskIT.kt
@@ -58,6 +58,62 @@ class GenerateApplicationPropertiesTaskIT {
 
     @Test
     @Order(2)
+    fun `should be able to auto generate a new property file when autoGenerateApplicationProperties is set to true`(
+        taskExecutor: TaskExecutor,
+        buildScript: BuildScript,
+        buildOutput: BuildOutput,
+    ) {
+        buildOutput
+            /** Clean up the previous output, so it won't interfere with the test */
+            .apply { listFiles()?.onEach(File::deleteRecursively) }
+            .resolve("resources/main/application.properties")
+            .shouldNotExist()
+
+        buildScript.asKts(
+            /** Coverage to test auto-generate functionality */
+            autoGenerateApplicationProperties = true,
+        )
+
+        /** An indirect run (i.e. via classes) */
+        taskExecutor
+            .execute("classes")
+            .outcome shouldBe Outcome.SUCCESS
+
+        buildOutput
+            .resolve("resources/main/application.properties")
+            .shouldExist()
+    }
+
+    @Test
+    @Order(3)
+    fun `should NOT be able to auto generate a new property file when autoGenerateApplicationProperties is set to false`(
+        taskExecutor: TaskExecutor,
+        buildScript: BuildScript,
+        buildOutput: BuildOutput,
+    ) {
+        buildOutput
+            /** Clean up the previous output, so it won't interfere with the test */
+            .apply { listFiles()?.onEach(File::deleteRecursively) }
+            .resolve("resources/main/application.properties")
+            .shouldNotExist()
+
+        buildScript.asKts(
+            /** Coverage to test auto-generate functionality */
+            autoGenerateApplicationProperties = false,
+        )
+
+        /** An indirect run (i.e. via classes) */
+        taskExecutor
+            .execute("classes")
+            .outcome shouldBe Outcome.SUCCESS
+
+        buildOutput
+            .resolve("resources/main/application.properties")
+            .shouldNotExist()
+    }
+
+    @Test
+    @Order(4)
     fun `should use cached output when task is ran and no sourcecode changes occurred when an explicit class compile dependency with matching source set`(
         taskExecutor: TaskExecutor,
         buildScript: BuildScript,
@@ -83,7 +139,7 @@ class GenerateApplicationPropertiesTaskIT {
     }
 
     @Test
-    @Order(3)
+    @Order(5)
     fun `should generate a new property file when the ApplicationProperties object has been modified`(
         taskExecutor: TaskExecutor,
         mainSourceSet: MainSourceSet,
@@ -112,7 +168,7 @@ class GenerateApplicationPropertiesTaskIT {
     }
 
     @Test
-    @Order(4)
+    @Order(6)
     fun `should generate a new property file when the Configuration extension has been modified`(
         taskExecutor: TaskExecutor,
         buildScript: BuildScript,
@@ -135,7 +191,7 @@ class GenerateApplicationPropertiesTaskIT {
     }
 
     @Test
-    @Order(5)
+    @Order(7)
     fun `should generate Propactive properties file for a class that is not located at root level`(
         taskExecutor: TaskExecutor,
         mainSourceSet: MainSourceSet,

--- a/propactive-plugin/src/test/kotlin/io/github/propactive/task/GenerateApplicationPropertiesTaskTest.kt
+++ b/propactive-plugin/src/test/kotlin/io/github/propactive/task/GenerateApplicationPropertiesTaskTest.kt
@@ -1,12 +1,9 @@
 package io.github.propactive.task
 
-import io.github.propactive.plugin.Configuration.Companion.DEFAULT_CLASS_COMPILE_DEPENDENCY
 import io.github.propactive.plugin.Configuration.Companion.DEFAULT_ENVIRONMENTS
 import io.github.propactive.plugin.Configuration.Companion.DEFAULT_IMPLEMENTATION_CLASS
 import io.github.propactive.plugin.Propactive.Companion.PROPACTIVE_GROUP
 import io.github.propactive.task.GenerateApplicationPropertiesTask.Companion.TASK_NAME
-import io.kotest.matchers.collections.shouldContain
-import io.kotest.matchers.collections.shouldNotBeEmpty
 import io.kotest.matchers.shouldBe
 import io.kotest.matchers.string.shouldBeBlank
 import io.kotest.matchers.string.shouldContain
@@ -29,12 +26,6 @@ class GenerateApplicationPropertiesTaskTest {
     @Test
     fun `should set task group`() {
         task.group shouldBe PROPACTIVE_GROUP
-    }
-
-    @Test
-    fun `should set class compile dependency`() {
-        task.dependsOn.shouldNotBeEmpty()
-        task.dependsOn shouldContain DEFAULT_CLASS_COMPILE_DEPENDENCY
     }
 
     @Test

--- a/propactive-plugin/src/test/kotlin/io/github/propactive/task/ValidateApplicationPropertiesTaskTest.kt
+++ b/propactive-plugin/src/test/kotlin/io/github/propactive/task/ValidateApplicationPropertiesTaskTest.kt
@@ -1,12 +1,9 @@
 package io.github.propactive.task
 
-import io.github.propactive.plugin.Configuration.Companion.DEFAULT_CLASS_COMPILE_DEPENDENCY
 import io.github.propactive.plugin.Configuration.Companion.DEFAULT_ENVIRONMENTS
 import io.github.propactive.plugin.Configuration.Companion.DEFAULT_IMPLEMENTATION_CLASS
 import io.github.propactive.plugin.Propactive.Companion.PROPACTIVE_GROUP
 import io.github.propactive.task.GenerateApplicationPropertiesTask.Companion.TASK_NAME
-import io.kotest.matchers.collections.shouldContain
-import io.kotest.matchers.collections.shouldNotBeEmpty
 import io.kotest.matchers.shouldBe
 import org.gradle.testfixtures.ProjectBuilder
 import org.junit.jupiter.api.BeforeEach
@@ -27,12 +24,6 @@ class ValidateApplicationPropertiesTaskTest {
     @Test
     fun `should set task group`() {
         task.group shouldBe PROPACTIVE_GROUP
-    }
-
-    @Test
-    fun `should set class compile dependency`() {
-        task.dependsOn.shouldNotBeEmpty()
-        task.dependsOn shouldContain DEFAULT_CLASS_COMPILE_DEPENDENCY
     }
 
     @Test

